cdc panic: runtime error: invalid memory address or nil pointer dereference

### What did you do?

This panic is seen after https://github.com/pingcap/tiflow/issues/10798

### What did you expect to see?

cdc no panic

### What did you see instead?


```
{"level":"warn","ts":"2024-03-17T03:44:29.64617Z","logger":"etcd-client","caller":"v3@v3.5.12/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0038b4000/upstream-pd-0.upstream-pd-peer.cdc-airbnb-simple-tian-gcp-tps-7500530-1-808.svc:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
{"level":"warn","ts":"2024-03-17T03:44:35.647313Z","logger":"etcd-client","caller":"v3@v3.5.12/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0038b4000/upstream-pd-0.upstream-pd-peer.cdc-airbnb-simple-tian-gcp-tps-7500530-1-808.svc:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
{"level":"warn","ts":"2024-03-17T03:44:41.648666Z","logger":"etcd-client","caller":"v3@v3.5.12/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0038b4000/upstream-pd-0.upstream-pd-peer.cdc-airbnb-simple-tian-gcp-tps-7500530-1-808.svc:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
{"level":"warn","ts":"2024-03-17T03:46:41.939417Z","logger":"etcd-client","caller":"v3@v3.5.12/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc003b921c0/upstream-pd-0.upstream-pd-peer.cdc-airbnb-simple-tian-gcp-tps-7500530-1-808.svc:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x3878de9]

goroutine 9470229 [running]:
github.com/pingcap/tiflow/cdc/kv.(*requestedStream).receive(0xc0706b43c0, {0x5c10998, 0xc0706b52c0}, 0xc0747ea580, 0xc06e0f2980, 0x1959945?, 0xc0706b48c0?)
        github.com/pingcap/tiflow/cdc/kv/shared_stream.go:225 +0x69
github.com/pingcap/tiflow/cdc/kv.(*requestedStream).run.func3()
        github.com/pingcap/tiflow/cdc/kv/shared_stream.go:187 +0x2e
golang.org/x/sync/errgroup.(*Group).Go.func1()
        golang.org/x/sync@v0.6.0/errgroup/errgroup.go:78 +0x56
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 9470087
        golang.org/x/sync@v0.6.0/errgroup/errgroup.go:75 +0x96
```

### Versions of the cluster

cdc version: 
cdc version:
Release Version: v8.0.0                                  
Git Commit Hash: 130403f3b9b8ad8a28ceada642277986e317ebc2
Git Branch: heads/refs/tags/v8.0.0                       
UTC Build Time: 2024-03-15 13:58:58                      
Go Version: go version go1.21.6 linux/amd64              
Failpoint Build: false

--- a/cdc/kv/shared_stream.go
+++ b/cdc/kv/shared_stream.go
@@ -169,7 +169,8 @@ func (s *requestedStream) run(ctx context.Context, c *SharedClient, rs *requeste
 		}
 	}()
 
-	cc, err := c.grpcPool.Connect(ctx, rs.storeAddr)
+	g, gctx := errgroup.WithContext(ctx)
+	cc, err := c.grpcPool.Connect(gctx, rs.storeAddr)
 	if err != nil {
 		log.Warn("event feed create grpc stream failed",
 			zap.String("namespace", c.changefeed.Namespace),
@@ -181,7 +182,6 @@ func (s *requestedStream) run(ctx context.Context, c *SharedClient, rs *requeste
 		return isCanceled()
 	}
 
-	g, gctx := errgroup.WithContext(ctx)
 	if cc.Multiplexing() {
 		s.multiplexing = cc
 		g.Go(func() error { return s.receive(gctx, c, rs, s.multiplexing, invalidSubscriptionID) })

--- a/cdc/kv/shared_stream.go
+++ b/cdc/kv/shared_stream.go
@@ -169,6 +169,8 @@ func (s *requestedStream) run(ctx context.Context, c *SharedClient, rs *requeste
 		}
 	}()
 
+	// grpc stream can be canceled by this context when any goroutine meet error,
+	// the underline established grpc connections is unaffected.
 	g, gctx := errgroup.WithContext(ctx)
 	cc, err := c.grpcPool.Connect(gctx, rs.storeAddr)
 	if err != nil {

--- a/cdc/kv/sharedconn/conn_and_client.go
+++ b/cdc/kv/sharedconn/conn_and_client.go
@@ -163,7 +163,6 @@ func (c *ConnAndClient) Multiplexing() bool {
 func (c *ConnAndClient) Release() {
 	if c.client != nil {
 		_ = c.client.CloseSend()
-		c.client = nil
 	}
 	if c.conn != nil && c.array != nil {
 		c.array.release(c.conn, false)

--- a/cdc/kv/sharedconn/conn_and_client.go
+++ b/cdc/kv/sharedconn/conn_and_client.go
@@ -56,6 +56,7 @@ type ConnAndClient struct {
 	conn   *Conn
 	array  *connArray
 	client cdcpb.ChangeData_EventFeedV2Client
+	closed atomic.Bool
 }
 
 // Conn is a connection.
@@ -161,8 +162,9 @@ func (c *ConnAndClient) Multiplexing() bool {
 
 // Release releases a ConnAndClient object.
 func (c *ConnAndClient) Release() {
-	if c.client != nil {
+	if c.client != nil && !c.closed.Load() {
 		_ = c.client.CloseSend()
+		c.closed.Store(true)
 	}
 	if c.conn != nil && c.array != nil {
 		c.array.release(c.conn, false)

--- a/tests/integration_tests/kafka_big_txn_v2/run.sh
+++ b/tests/integration_tests/kafka_big_txn_v2/run.sh
@@ -45,12 +45,9 @@ function run() {
 	check_table_exists "big_txn.usertable" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT}
 	run_sql "CREATE TABLE big_txn.usertable1 LIKE big_txn.usertable" ${UP_TIDB_HOST} ${UP_TIDB_PORT}
 	run_sql "INSERT INTO big_txn.usertable1 SELECT * FROM big_txn.usertable" ${UP_TIDB_HOST} ${UP_TIDB_PORT}
-	sleep 60
-	check_table_exists "big_txn.usertable1" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT}
-
-	run_sql "CREATE TABLE big_txn.finish_mark_1 (a int primary key);"
+	run_sql "CREATE TABLE big_txn.finish_mark (a int primary key);"
 	sleep 120
-	check_table_exists "big_txn.finish_mark_1" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT} 60
+	check_table_exists "big_txn.finish_mark" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT} 120
 
 	check_sync_diff $WORK_DIR $CUR/conf/diff_config.toml
 

